File: Backend/src/services/mcp.service.js
const axios = require("axios");

const MCP_URL = "https://api.githubcopilot.com/mcp/";

const mcpClient = axios.create({
  baseURL: MCP_URL,
  timeout: 30000,
  headers: {
    Authorization: `Bearer ${process.env.GITHUB_MCP_PAT}`,
    "Content-Type": "application/json",
  },
});

const parseMcpResponse = (rawResponse) => {
  try {
    if (!rawResponse) {
      throw new Error("Empty MCP response");
    }

    if (typeof rawResponse === "object") {
      return rawResponse;
    }

    const jsonText = rawResponse
      .replace("event: message", "")
      .replace("data:", "")
      .trim();

    return JSON.parse(jsonText);
  } catch (error) {
    console.error("MCP Response Parse Error:", error.message);

    return {
      error: true,
      message: "Unable to parse MCP response.",
      raw: rawResponse,
    };
  }
};

const executeMcpTool = async (
  toolName,
  args = {},
) => {
  try {
    if (!process.env.GITHUB_MCP_PAT) {
      throw new Error(
        "GITHUB_MCP_PAT is missing.",
      );
    }

    const response = await mcpClient.post("", {
      jsonrpc: "2.0",
      id: 1,
      method: "tools/call",
      params: {
        name: toolName,
        arguments: args,
      },
    });

    const parsedResponse = parseMcpResponse(
      response.data,
    );

    if (parsedResponse?.error) {
      throw new Error(
        parsedResponse.message ||
          "Tool execution failed.",
      );
    }

    return parsedResponse;
  } catch (error) {
    console.error("\n========== MCP ERROR ==========\n");

    console.error(
      error.response?.data || error.message,
    );

    console.error(
      "\n===============================\n",
    );

    throw new Error(
      error.response?.data?.message ||
        error.message ||
        "Failed to execute MCP tool."
    );
  }
};

const getAvailableTools = async () => {
  try {
    if (!process.env.GITHUB_MCP_PAT) {
      throw new Error(
        "GITHUB_MCP_PAT is missing."
      );
    }

    const response = await mcpClient.post("", {
      jsonrpc: "2.0",
      id: 1,
      method: "tools/list",
    });

    const parsedResponse = parseMcpResponse(
      response.data,
    );

    const totalTools =
      parsedResponse?.result?.tools?.length || 0;

    console.log("\n========== TOOLS RESPONSE ==========\n");

    console.log(
      `Loaded ${totalTools} tools`
    );

    console.log(
      "\n====================================\n",
    );

    return parsedResponse;
  } catch (error) {
    console.error(
      "\n========== TOOLS ERROR ==========\n",
    );

    console.error(
      error.response?.data || error.message,
    );

    console.error(
      "\n=================================\n",
    );

    return null;
  }
};

module.exports = {
  executeMcpTool,
  getAvailableTools,
};

File: Backend/app/services/mcp_service.py
"""
MCP service — handles communication with the GitHub Remote MCP Server.
Port of services/mcp.service.js.
"""

import json
import logging

from app.clients.mcp_client import mcp_request
from app.core.config import get_settings

logger = logging.getLogger(__name__)


def parse_mcp_response(raw_response) -> dict:
    """
    Parse an MCP response which may be either a JSON object
    or an SSE-formatted string with 'event: message\\ndata: {...}'.

    Exact port of parseMcpResponse() from mcp.service.js.

    Args:
        raw_response: The raw MCP response (dict or string).

    Returns:
        Parsed dict.
    """
    try:
        if not raw_response:
            raise ValueError("Empty MCP response")

        if isinstance(raw_response, dict):
            return raw_response

        # Handle SSE format safely
        text = str(raw_response).strip()
        if text.startswith("event:"):
            for line in text.splitlines():
                if line.startswith("data: "):
                    return json.loads(line[6:])
                elif line.startswith("data:"):
                    return json.loads(line[5:])
        
        return json.loads(text)
    except Exception as e:
        logger.error("MCP Response Parse Error: %s", str(e))
        return {
            "error": True,
            "message": "Unable to parse MCP response.",
            "raw": str(raw_response),
        }


async def execute_mcp_tool(tool_name: str, args: dict | None = None) -> dict:
    """
    Execute an MCP tool via the Remote MCP Server.

    Args:
        tool_name: The name of the MCP tool to execute.
        args: Arguments for the tool.

    Returns:
        Parsed MCP response.

    Raises:
        Exception: On MCP errors.
    """
    try:
        settings = get_settings()
        if not settings.GITHUB_MCP_PAT:
            raise RuntimeError("GITHUB_MCP_PAT is missing.")

        response = await mcp_request(
            method="tools/call",
            params={
                "name": tool_name,
                "arguments": args or {},
            },
        )

        parsed = parse_mcp_response(response)

        if parsed.get("error"):
            raise RuntimeError(
                parsed.get("message", "Tool execution failed.")
            )

        return parsed

    except Exception as error:
        logger.error("\n========== MCP ERROR ==========\n")
        logger.error(str(error))
        logger.error("\n===============================\n")

        raise RuntimeError(
            str(error) or "Failed to execute MCP tool."
        )


async def get_available_tools() -> dict | None:
    """
    Fetch the list of available MCP tools from the Remote MCP Server.

    Returns:
        Parsed MCP response with result.tools, or None on failure.
    """
    try:
        settings = get_settings()
        if not settings.GITHUB_MCP_PAT:
            raise RuntimeError("GITHUB_MCP_PAT is missing.")

        response = await mcp_request(method="tools/list")

        parsed = parse_mcp_response(response)

        tools = parsed.get("result", {}).get("tools", [])
        total_tools = len(tools) if isinstance(tools, list) else 0

        logger.info("\n========== TOOLS RESPONSE ==========\n")
        logger.info("Loaded %d tools", total_tools)
        logger.info("\n====================================\n")

        return parsed

    except Exception as error:
        logger.error("\n========== TOOLS ERROR ==========\n")
        logger.error(str(error))
        logger.error("\n=================================\n")
        return None
